# `systems data sources`

&nbsp;

###### [Boris Steipe](https://orcid.org/0000-0002-1134-6758),
###### Department of Biochemistry and Department of Molecular Genetics,
###### University of Toronto
###### Canada
###### &lt;boris.steipe@utoronto.ca&gt;

----

A collection of data sources that are relevant for computational systems biology.

----

&nbsp;

### System

**A [system](http://steipe.biochemistry.utoronto.ca/abc/assets/BIN-SYS-Concepts.pdf) maps a set of collaborating components to its emergent behaviour.**

"System" is a concept. In order to understand biology, we apply the concept to observables.

* We identify *components*,
* we collect evidence of their *collaboration*,
* we analyze the system's *behaviour* and specify what is *emergent* about it,
* we organize our knowledge in a *system architecture model*.

**A system architecture describes the system's behaviour in terms of its subsystems and their relationships, given its context, within its boundaries.**

The observables in this analysis-by-modelling are:

* molecular components and sets of components
* their relationships
* the functional consequences

&nbsp;

### Data resources

&nbsp;

##### Resource collections
* **[The NAR Database Issue (2019)](https://academic.oup.com/nar/issue/47/D1)**
* **[The NAR Web Server Issue (2018)](https://academic.oup.com/nar/issue/46/W1)**

&nbsp;

##### Molecular components

**Identifiers**

* `HGNC.RData`: reference symbols and identifiers for human genes.
* `HGNCdata.R`: The script that produced `HGNC.RData`

&nbsp;

**Annotations**

* **[UCSC](https://genome.ucsc.edu/)**
* **[Conservation (EggNOG)](http://eggnogdb.embl.de/#/app/home)**
* **[Orthologues (OrthoDB)](https://www.orthodb.org/)**
* **[Gene Ontology Annotations](https://www.ebi.ac.uk/GOA)**
* **[SMART database](http://smart.embl-heidelberg.de/)**
* **[Pfam database](https://pfam.xfam.org/)**
* **[STRING](https://string-db.org/)**
* **[ENCODE](https://www.encodeproject.org/)**
* **[Human Protein Atlas](https://www.proteinatlas.org/)**
* **[PDB](https://www.rcsb.org/)**

**Dynamics**

* **[GEO expression profiles](https://www.ncbi.nlm.nih.gov/geo/)**
* **[iProx Proteome data](http://www.iprox.org/)**

&nbsp;

##### Relationships (networks)

* **[STRING](https://string-db.org/)**
* **[IntAct](https://www.ebi.ac.uk/intact/)**
* **[iRefIndex](http://irefindex.org/)**
* **[BioGRID](https://thebiogrid.org/)**
* **[IntAct](https://www.ebi.ac.uk/intact/)**
* **[GO: semantic similarity (from Bioconductor tools)](https://www.bioconductor.org/help/search/index.html?q=%22semantic+similarity%22/)**
* phylogenetic profiling (?)
* **[CoExpress DB](http://coxpresdb.jp/)**
* **[CoExpedia](http://www.coexpedia.org/)**
* **[ENPD (Eukaryotic Nucleic Acid Binding Proteins)](http://qinlab.sls.cuhk.edu.hk/ENPD/)**

&nbsp;

##### Modules

* **[MSigDB](http://software.broadinstitute.org/gsea/msigdb)**
* **[GeneMania](https://genemania.org/)**
* **[Reactome Pathway Browser](https://reactome.org/PathwayBrowser/#/)**
* **[Wikipathways](https://www.wikipathways.org/)**
* **[KEGG pathway database](https://www.genome.jp/kegg/pathway.html)**
* **[PANTHER protein annotation](https://www.nature.com/articles/nprot.2013.092)**
* **[NCI-Nature Pathway Interaction Database](http://pid.nci.nih.gov/)**
* **[Pathway Commons](http://www.pathwaycommons.org/)**
* **[Consensus Path DB(CPDB)](http://consensuspathdb.org/)**
* **[hu.MAP Protein complexes](http://proteincomplexes.org/)**
* **[CORUM](http://mips.helmholtz-muenchen.de/corum/)**
* **[GWAS catalog](https://www.ebi.ac.uk/gwas/)**


&nbsp;

##### Function and phenotypes

* **[OMIM](https://www.omim.org/)**
* **[Human Disease Ontology](https://www.ebi.ac.uk/ols/ontologies/doid)**
* **[Human Phenotype Ontology](https://hpo.jax.org/app/)**
* **[COSMIC (Catalogue of Somatic Mutations in Cancer)](https://cancer.sanger.ac.uk/cosmic)**
* **[Navigome](https://omictools.com/navigome-tool)**
* **[DAVID](https://david.ncifcrf.gov/)**
* **[Bystro](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5801807/)**


&nbsp;

<!-- END -->
